File: app/src/main/java/com/example/popularmovies/data/MoviesProvider.java
package com.example.popularmovies.data;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteQueryBuilder;
import android.net.Uri;
import android.provider.BaseColumns;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * A content provider for movies and list of movies
 * Created by rafael on 01/07/16.
 */
public class MoviesProvider extends ContentProvider {
    /**
     * Constant for list requests
     */
    public static final int LIST = 100;

    /**
     * Constant for movie list requests (ordered by popularity by default)
     */
    public static final int LIST_BY_ID = 101;

    public static final int LIST_BY_SELECTION = 102;

    /**
     * Constant for specific movie requests
     */
    public static final int MOVIE = 200;

    /**
     * Constant for specific movie requests
     */
    public static final int MOVIE_BY_ID = 201;

    public static final int MOVIES_BY_SELECTION = 202;


    public static final int MOVIE_LIST = 300;

    public static final int MOVIE_LIST_BY_ID = 301;

    public static final int MOVIE_LIST_BY_SELECTION = 302;

    /**
     * Static UriMatcher for the provider
     */
    private static final UriMatcher sUriMatcher = buildUriMatcher();

    private MoviesDbHelper mDbOpener;

    private static final String BY_ID_SELECTION = BaseColumns._ID + "=?";

    private static final String LIST_SELECTION = MoviesContract.ListEntry.COLUMN_SELECTION + "=?";

    private static final String BY_LIST_SELECTION_SELECTION = MoviesContract.MovieListEntry.COLUMN_LIST_KEY +
            " IN (SELECT " + MoviesContract.ListEntry.TABLE_NAME + "." + MoviesContract.ListEntry._ID +
            " FROM " + MoviesContract.ListEntry.TABLE_NAME + " WHERE " + MoviesContract.ListEntry.TABLE_NAME +
            "." + MoviesContract.ListEntry.COLUMN_SELECTION + "=?)";

    private static final SQLiteQueryBuilder sMoviesByListQueryBuilder = new SQLiteQueryBuilder();

    private static final String sMovieListSelectionByOrder =
            MoviesContract.MovieListEntry.COLUMN_LIST_KEY + "=?";

    private static final String sMovieOrder = MoviesContract.MovieEntry.COLUMN_VOTE_AVG + " DESC";

    static {
        sMoviesByListQueryBuilder.setTables(
                MoviesContract.MovieEntry.TABLE_NAME + " INNER JOIN " +
                MoviesContract.MovieListEntry.TABLE_NAME + " ON " +
                MoviesContract.MovieEntry._ID + "=" + MoviesContract.MovieListEntry.COLUMN_MOVIE_KEY +
                " INNER JOIN " + MoviesContract.ListEntry.TABLE_NAME + " ON " +
                MoviesContract.ListEntry._ID + "=" + MoviesContract.MovieListEntry.COLUMN_LIST_KEY
        );
    }

    /**
     * Builds an UriMatcher for the provider
     * @return An UriMatcher
     */
    public static UriMatcher buildUriMatcher() {
        final String authority = MoviesContract.CONTENT_AUTHORITY;
        UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION, LIST);
        uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION + "/#", LIST_BY_ID);
        uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION + "/*", LIST_BY_SELECTION);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION, MOVIE);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION + "/#", MOVIE_BY_ID);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION + "/*", MOVIES_BY_SELECTION);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION, MOVIE_LIST);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION + "/#", MOVIE_LIST_BY_ID);
        uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION + "/*", MOVIE_LIST_BY_SELECTION);

        return uriMatcher;
    }

    @Override
    public boolean onCreate() {
        mDbOpener = new MoviesDbHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        Cursor retCursor;
        SQLiteDatabase db = mDbOpener.getReadableDatabase();

        int match = sUriMatcher.match(uri);
        switch(match) {
            case LIST:
                retCursor = db.query(
                        MoviesContract.ListEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder
                );
                break;
            case LIST_BY_ID:
                Long listId = MoviesContract.ListEntry.getIdFromUri(uri);
                retCursor = db.query(
                        MoviesContract.ListEntry.TABLE_NAME,
                        projection,
                        BY_ID_SELECTION,
                        new String[] { Long.toString(listId) },
                        null,
                        null,
                        sortOrder
                );
                break;
            case LIST_BY_SELECTION:
                String listSelection = MoviesContract.ListEntry.getSelectionFromUri(uri);
                retCursor = db.query(
                        MoviesContract.ListEntry.TABLE_NAME,
                        projection,
                        LIST_SELECTION,
                        new String[] { listSelection },
                        null,
                        null,
                        sortOrder
                );
                break;
            case MOVIE:
                retCursor = db.query(
                        MoviesContract.MovieEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder
                );
                break;
            case MOVIE_BY_ID:
                retCursor = getMovieById(uri, projection);
                break;
            case MOVIE_LIST:
                retCursor = db.query(
                        MoviesContract.MovieListEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder
                );
                break;
            case MOVIE_LIST_BY_ID:
                Long movieListId = MoviesContract.MovieListEntry.getIdFromUri(uri);
                retCursor = db.query(
                        MoviesContract.MovieListEntry.TABLE_NAME,
                        projection,
                        BY_ID_SELECTION,
                        new String[] { Long.toString(movieListId) },
                        null,
                        null,
                        sortOrder
                );
                break;
            case MOVIE_LIST_BY_SELECTION:
                listSelection = MoviesContract.MovieListEntry.getSelectionFromUri(uri);
                retCursor = db.query(
                        MoviesContract.MovieListEntry.TABLE_NAME,
                        projection,
                        BY_LIST_SELECTION_SELECTION,
                        new String[] { listSelection },
                        null,
                        null,
                        sortOrder
                );
                break;
            default:
                throw new UnsupportedOperationException("Unknown query uri: " + uri);
        }
        retCursor.setNotificationUri(getContext().getContentResolver(), uri);
        return retCursor;
    }

    @Override
    public String getType(@NonNull Uri uri) {
        int match = sUriMatcher.match(uri);

        switch (match) {
            case LIST:
                return MoviesContract.ListEntry.CONTENT_TYPE;
            case LIST_BY_ID:
                return MoviesContract.ListEntry.CONTENT_ITEM_TYPE;
            case LIST_BY_SELECTION:
                return MoviesContract.ListEntry.CONTENT_ITEM_TYPE;
            case MOVIE:
                return MoviesContract.MovieEntry.CONTENT_TYPE;
            case MOVIE_BY_ID:
                return MoviesContract.MovieEntry.CONTENT_ITEM_TYPE;
            case MOVIES_BY_SELECTION:
                return MoviesContract.MovieEntry.CONTENT_TYPE;
            case MOVIE_LIST:
                return MoviesContract.MovieListEntry.CONTENT_TYPE;
            case MOVIE_LIST_BY_ID:
                return MoviesContract.MovieListEntry.CONTENT_ITEM_TYPE;
            case MOVIE_LIST_BY_SELECTION:
                return MoviesContract.MovieListEntry.CONTENT_TYPE;
            default:
                throw new UnsupportedOperationException("Unknown content URI: " + uri);
        }
    }

    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        final SQLiteDatabase database = mDbOpener.getWritableDatabase();
        final int match = sUriMatcher.match(uri);
        Uri returnUri;
        switch (match) {
            case LIST:
                Long _id = database.insert(MoviesContract.ListEntry.TABLE_NAME, null, values);
                if (_id <= 0) {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                returnUri = MoviesContract.ListEntry.buildUri(_id);
                break;
            case MOVIE:
                _id = database.insert(MoviesContract.MovieEntry.TABLE_NAME, null, values);
                if (_id <= 0) {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                returnUri = MoviesContract.MovieEntry.buildUri(_id);
                break;
            case MOVIE_LIST:
                _id = database.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, values);
                if (_id <= 0) {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                returnUri = MoviesContract.MovieListEntry.buildUri(_id);
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        return returnUri;
    }

    @Override
    public int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {
        final SQLiteDatabase db = mDbOpener.getWritableDatabase();
        final int match = sUriMatcher.match(uri);
        int rowsDeleted;
        // this makes delete all rows return the number of rows deleted
        if (null == selection) selection = "1";
        switch (match) {
            case LIST:
                rowsDeleted = db.delete(MoviesContract.ListEntry.TABLE_NAME, selection, selectionArgs);
                break;
            case MOVIE:
                rowsDeleted = db.delete(MoviesContract.MovieEntry.TABLE_NAME, selection, selectionArgs);
                break;
            case MOVIE_LIST:
                rowsDeleted = db.delete(MoviesContract.MovieListEntry.TABLE_NAME, selection, selectionArgs);
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        if (rowsDeleted > 0) {
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return rowsDeleted;
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        final SQLiteDatabase db = mDbOpener.getWritableDatabase();
        final int match = sUriMatcher.match(uri);
        int rowsUpdated;

        switch (match) {
            case LIST_BY_ID:
                Long listId = MoviesContract.ListEntry.getIdFromUri(uri);
                rowsUpdated = db.update(
                        MoviesContract.ListEntry.TABLE_NAME,
                        values,
                        BY_ID_SELECTION,
                        new String[] { Long.toString(listId) }
                );
                break;
            case MOVIE_BY_ID:
                Long movieId = MoviesContract.MovieEntry.getIdFromUri(uri);
                rowsUpdated = db.update(
                        MoviesContract.MovieEntry.TABLE_NAME,
                        values,
                        BY_ID_SELECTION,
                        new String[] { Long.toString(movieId) }
                );
                break;
            default:
                throw new UnsupportedOperationException("Unknown update uri: " + uri);
        }
        if (rowsUpdated > 0) {
            getContext().getContentResolver().notifyChange(uri, null);
        }

        return rowsUpdated;
    }

    private Cursor getMovieById(Uri uri, String[] projection) {
        Long id = MoviesContract.MovieEntry.getIdFromUri(uri);
        return mDbOpener.getReadableDatabase().query(
                MoviesContract.MovieEntry.TABLE_NAME,
                projection,
                BY_ID_SELECTION,
                new String[] { Long.toString(id) },
                null,
                null,
                null
        );
    }

    private Cursor getMovies(String[] projection, @Nullable String listSelector, @Nullable  String sortOrder) {
        SQLiteDatabase db = mDbOpener.getReadableDatabase();
        if (null == listSelector) listSelector = MoviesContract.ListSelections.MOST_POPULAR.toString();
        if (null == sortOrder) sortOrder = sMovieOrder;
        return sMoviesByListQueryBuilder.query(
                db,
                projection,
                BY_ID_SELECTION,
                new String[] { listSelector },
                null,
                null,
                sortOrder
        );
    }

    @Override
    public int bulkInsert(@NonNull Uri uri, @NonNull ContentValues[] values) {
        final int match = sUriMatcher.match(uri);
        SQLiteDatabase db = mDbOpener.getWritableDatabase();
        int rowsDeleted, rowsInserted = 0;
        switch (match) {
            case MOVIE_LIST_BY_SELECTION:
                rowsDeleted = clearMovieListBySelection(uri);
                for (ContentValues movieList : values) {
                    Long movieListId = db.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, movieList);
                    if (movieListId > 0) rowsInserted++;
                }
                clearOrphanedMovies();
                break;
            default:
                throw new UnsupportedOperationException("Unknown bulk insert uri: " + uri);
        }
        if (rowsDeleted + rowsInserted > 0) {
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return rowsInserted;
    }

    private Integer clearMovieListBySelection(final Uri listBySelectionUri) {
        Cursor listCursor = query(listBySelectionUri, null, null, null, null);
        if (!listCursor.moveToFirst()) {
            listCursor.close();
            return 0;
        }
        Integer idIndex = listCursor.getColumnIndex(MoviesContract.ListEntry._ID);
        Long listId = listCursor.getLong(idIndex);
        listCursor.close();
        SQLiteDatabase db = mDbOpener.getWritableDatabase();
        final String byListIdSelection = MoviesContract.MovieListEntry.COLUMN_LIST_KEY + "=?";
        return db.delete(
                MoviesContract.MovieListEntry.TABLE_NAME,
                byListIdSelection,
                new String[] { Long.toString(listId) }
        );
    }

    private Integer clearOrphanedMovies() {
        SQLiteDatabase db = mDbOpener.getWritableDatabase();
        final String orphanedSelection = MoviesContract.MovieEntry._ID + " NOT IN (SELECT " +
                MoviesContract.MovieListEntry.TABLE_NAME + "." + MoviesContract.MovieListEntry.COLUMN_MOVIE_KEY +
                " FROM " + MoviesContract.MovieListEntry.TABLE_NAME + ")";
        return db.delete(MoviesContract.MovieEntry.TABLE_NAME, orphanedSelection, null);
    }
}

File: app/src/main/java/com/example/popularmovies/data/MoviesProvider.kt
package com.example.popularmovies.data

import android.content.ContentProvider
import android.content.ContentResolver
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteQueryBuilder
import android.net.Uri
import android.provider.BaseColumns

/**
 * A content provider for movies and list of movies
 * Created by rafael on 01/07/16.
 */
class MoviesProvider : ContentProvider() {

    private var mDbOpener: MoviesDbHelper? = null

    override fun onCreate(): Boolean {
        mDbOpener = MoviesDbHelper(context!!)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val db = mDbOpener?.readableDatabase ?: return null

        val match = sUriMatcher.match(uri)
        val retCursor: Cursor? = when (match) {
            LIST -> db.query(
                    MoviesContract.ListEntry.TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs, null, null,
                    sortOrder
            )
            LIST_BY_ID -> {
                val listId = MoviesContract.ListEntry.getIdFromUri(uri)
                db.query(
                        MoviesContract.ListEntry.TABLE_NAME,
                        projection,
                        BY_ID_SELECTION,
                        arrayOf(java.lang.Long.toString(listId)), null, null,
                        sortOrder
                )
            }
            LIST_BY_SELECTION -> {
                val listSelection = MoviesContract.ListEntry.getSelectionFromUri(uri)
                db.query(
                        MoviesContract.ListEntry.TABLE_NAME,
                        projection,
                        LIST_SELECTION,
                        arrayOf(listSelection), null, null,
                        sortOrder
                )
            }
            MOVIE -> db.query(
                    MoviesContract.MovieEntry.TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs, null, null,
                    sortOrder
            )
            MOVIE_BY_ID -> getMovieById(uri, projection)
            MOVIES_BY_SELECTION -> {
                val listSelection = MoviesContract.MovieEntry.getSelectionFromUri(uri)
                getMovies(projection, listSelection, sortOrder)
            }
            MOVIE_LIST -> db.query(
                    MoviesContract.MovieListEntry.TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs, null, null,
                    sortOrder
            )
            MOVIE_LIST_BY_ID -> {
                val movieListId = MoviesContract.MovieListEntry.getIdFromUri(uri)
                db.query(
                        MoviesContract.MovieListEntry.TABLE_NAME,
                        projection,
                        BY_ID_SELECTION,
                        arrayOf(java.lang.Long.toString(movieListId!!)), null, null,
                        sortOrder
                )
            }
            MOVIE_LIST_BY_SELECTION -> {
                val listSelection = MoviesContract.MovieListEntry.getSelectionFromUri(uri)
                db.query(
                        MoviesContract.MovieListEntry.TABLE_NAME,
                        projection,
                        BY_LIST_SELECTION_SELECTION,
                        arrayOf(listSelection), null, null,
                        sortOrder
                )
            }
            else -> throw UnsupportedOperationException("Unknown query uri: " + uri)
        }
        val contentResolver = context?.contentResolver
        if (contentResolver != null) {
            retCursor?.setNotificationUri(contentResolver, uri)
        }
        return retCursor
    }

    override fun getType(uri: Uri): String {
        val match = sUriMatcher.match(uri)

        return when (match) {
            LIST -> MoviesContract.ListEntry.CONTENT_TYPE
            LIST_BY_ID -> MoviesContract.ListEntry.CONTENT_ITEM_TYPE
            LIST_BY_SELECTION -> MoviesContract.ListEntry.CONTENT_ITEM_TYPE
            MOVIE -> MoviesContract.MovieEntry.CONTENT_TYPE
            MOVIE_BY_ID -> MoviesContract.MovieEntry.CONTENT_ITEM_TYPE
            MOVIES_BY_SELECTION -> MoviesContract.MovieEntry.CONTENT_TYPE
            MOVIE_LIST -> MoviesContract.MovieListEntry.CONTENT_TYPE
            MOVIE_LIST_BY_ID -> MoviesContract.MovieListEntry.CONTENT_ITEM_TYPE
            MOVIE_LIST_BY_SELECTION -> MoviesContract.MovieListEntry.CONTENT_TYPE
            else -> throw UnsupportedOperationException("Unknown content URI: " + uri)
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val database = mDbOpener?.writableDatabase ?: return null
        val match = sUriMatcher.match(uri)
        return when (match) {
            LIST -> {
                val _id = database.insert(MoviesContract.ListEntry.TABLE_NAME, null, values)
                if (_id <= 0) {
                    throw SQLException("Failed to insert row into " + uri)
                }
                MoviesContract.ListEntry.buildUri(_id)
            }
            MOVIE -> {
                val _id = database.insert(MoviesContract.MovieEntry.TABLE_NAME, null, values)
                if (_id <= 0) {
                    throw SQLException("Failed to insert row into " + uri)
                }
                MoviesContract.MovieEntry.buildUri(_id)
            }
            MOVIE_LIST -> {
                val _id = database.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, values)
                if (_id <= 0) {
                    throw SQLException("Failed to insert row into " + uri)
                }
                MoviesContract.MovieListEntry.buildUri(_id)
            }
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        var selection = selection
        val db = mDbOpener?.writableDatabase ?: return -1
        val match = sUriMatcher.match(uri)
        // this makes delete all rows return the number of rows deleted
        if (null == selection) selection = "1"
        val rowsDeleted: Int = when (match) {
            LIST -> db.delete(MoviesContract.ListEntry.TABLE_NAME, selection, selectionArgs)
            MOVIE -> db.delete(MoviesContract.MovieEntry.TABLE_NAME, selection, selectionArgs)
            MOVIE_LIST -> db.delete(MoviesContract.MovieListEntry.TABLE_NAME, selection, selectionArgs)
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }
        val contentResolver = context?.contentResolver
        if (rowsDeleted > 0 && contentResolver != null) {
            contentResolver.notifyChange(uri, null)
        }
        return rowsDeleted
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        val db = mDbOpener?.writableDatabase ?: return -1
        val match = sUriMatcher.match(uri)
        val rowsUpdated: Int = when (match) {
            LIST_BY_ID -> {
                val listId = MoviesContract.ListEntry.getIdFromUri(uri)
                db.update(
                        MoviesContract.ListEntry.TABLE_NAME,
                        values,
                        BY_ID_SELECTION,
                        arrayOf(java.lang.Long.toString(listId))
                )
            }
            MOVIE_BY_ID -> {
                val movieId = MoviesContract.MovieEntry.getIdFromUri(uri)
                db.update(
                        MoviesContract.MovieEntry.TABLE_NAME,
                        values,
                        BY_ID_SELECTION,
                        arrayOf(java.lang.Long.toString(movieId!!))
                )
            }
            else -> throw UnsupportedOperationException("Unknown update uri: " + uri)
        }
        val contentResolver = context?.contentResolver
        if (rowsUpdated > 0 && contentResolver != null) {
            contentResolver.notifyChange(uri, null)
        }

        return rowsUpdated
    }

    private fun getMovieById(uri: Uri, projection: Array<String>?): Cursor? {
        val id = MoviesContract.MovieEntry.getIdFromUri(uri)
        return mDbOpener?.readableDatabase?.query(
                MoviesContract.MovieEntry.TABLE_NAME,
                projection,
                BY_ID_SELECTION,
                arrayOf(java.lang.Long.toString(id)), null, null, null
        )
    }

    private fun getMovies(projection: Array<String>?, listSelector: String?, sortOrder: String?): Cursor? {
        var listSelector = listSelector
        var sortOrder = sortOrder
        val db = mDbOpener?.readableDatabase ?: return null
        if (null == listSelector) {
            listSelector = MoviesContract.ListSelections.MOST_POPULAR.toString()
        }
        if (null == sortOrder) {
            sortOrder = sMovieOrder
        }
        return sMoviesByListQueryBuilder.query(
                db,
                projection,
                LIST_SELECTION,
                arrayOf(listSelector), null, null,
                sortOrder
        )
    }

    override fun bulkInsert(uri: Uri, values: Array<ContentValues>): Int {
        val match = sUriMatcher.match(uri)
        val db = mDbOpener?.writableDatabase ?: return -1
        val rowsDeleted: Int
        var rowsInserted = 0
        when (match) {
            MOVIE_LIST_BY_SELECTION -> {
                val listSelection = MoviesContract.MovieListEntry.getSelectionFromUri(uri)
                val listByIdUri = MoviesContract.ListEntry.buildBySelectionUri(listSelection)
                rowsDeleted = clearMovieListBySelection(listByIdUri)
                values
                        .map { db.insert(MoviesContract.MovieListEntry.TABLE_NAME, null, it) }
                        .filter { it > 0 }
                        .forEach { rowsInserted++ }
                clearOrphanedMovies()
            }
            else -> throw UnsupportedOperationException("Unknown bulk insert uri: " + uri)
        }
        val contentResolver = context?.contentResolver
        if (rowsDeleted + rowsInserted > 0 && contentResolver != null) {
            contentResolver.notifyChange(uri, null)
        }
        return rowsInserted
    }

    private fun clearMovieListBySelection(listBySelectionUri: Uri): Int {
        val listCursor = query(listBySelectionUri, null, null, null, null) ?: return 0
        if (!listCursor.moveToFirst()) {
            listCursor.close()
            return 0
        }
        val idIndex = listCursor.getColumnIndex(MoviesContract.ListEntry._ID)
        val listId = listCursor.getLong(idIndex)
        listCursor.close()
        val db = mDbOpener?.writableDatabase ?: return 0
        val byListIdSelection = MoviesContract.MovieListEntry.COLUMN_LIST_KEY + "=?"
        return db.delete(
                MoviesContract.MovieListEntry.TABLE_NAME,
                byListIdSelection,
                arrayOf(java.lang.Long.toString(listId))
        )
    }

    private fun clearOrphanedMovies(): Int {
        val db = mDbOpener?.writableDatabase ?: return 0
        val orphanedSelection = MoviesContract.MovieEntry._ID + " NOT IN (SELECT " +
                MoviesContract.MovieListEntry.TABLE_NAME + "." + MoviesContract.MovieListEntry.COLUMN_MOVIE_KEY +
                " FROM " + MoviesContract.MovieListEntry.TABLE_NAME + ")"
        return db.delete(MoviesContract.MovieEntry.TABLE_NAME, orphanedSelection, null)
    }

    companion object {
        /**
         * Constant for list requests
         */
        val LIST = 100

        /**
         * Constant for movie list requests (ordered by popularity by default)
         */
        val LIST_BY_ID = 101

        val LIST_BY_SELECTION = 102

        /**
         * Constant for specific movie requests
         */
        val MOVIE = 200

        /**
         * Constant for specific movie requests
         */
        val MOVIE_BY_ID = 201

        val MOVIES_BY_SELECTION = 202


        val MOVIE_LIST = 300

        val MOVIE_LIST_BY_ID = 301

        val MOVIE_LIST_BY_SELECTION = 302

        /**
         * Static UriMatcher for the provider
         */
        private val sUriMatcher = buildUriMatcher()

        private val BY_ID_SELECTION = BaseColumns._ID + "=?"

        private val LIST_SELECTION = MoviesContract.ListEntry.TABLE_NAME + "." +
                MoviesContract.ListEntry.COLUMN_SELECTION + "=?"

        private val BY_LIST_SELECTION_SELECTION = MoviesContract.MovieListEntry.COLUMN_LIST_KEY +
                " IN (SELECT " + MoviesContract.ListEntry.TABLE_NAME + "." + MoviesContract.ListEntry._ID +
                " FROM " + MoviesContract.ListEntry.TABLE_NAME + " WHERE " + MoviesContract.ListEntry.TABLE_NAME +
                "." + MoviesContract.ListEntry.COLUMN_SELECTION + "=?)"

        private val sMovieOrder = MoviesContract.MovieEntry.COLUMN_VOTE_AVG + " DESC"

        private val sMoviesByListQueryBuilder = SQLiteQueryBuilder()

        init {
            sMoviesByListQueryBuilder.tables = MoviesContract.MovieEntry.TABLE_NAME + " INNER JOIN " +
                    MoviesContract.MovieListEntry.TABLE_NAME + " ON " +
                    MoviesContract.MovieEntry.TABLE_NAME + "." + MoviesContract.MovieEntry._ID + "=" +
                    MoviesContract.MovieListEntry.TABLE_NAME + "." + MoviesContract.MovieListEntry.COLUMN_MOVIE_KEY +
                    " INNER JOIN " + MoviesContract.ListEntry.TABLE_NAME + " ON " +
                    MoviesContract.ListEntry.TABLE_NAME + "." + MoviesContract.ListEntry._ID + "=" +
                    MoviesContract.MovieListEntry.TABLE_NAME + "." + MoviesContract.MovieListEntry.COLUMN_LIST_KEY
        }

        /**
         * Builds an UriMatcher for the provider
         * @return An UriMatcher
         */
        fun buildUriMatcher(): UriMatcher {
            val authority = MoviesContract.CONTENT_AUTHORITY
            val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
            uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION, LIST)
            uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION + "/#", LIST_BY_ID)
            uriMatcher.addURI(authority, MoviesContract.LIST_LOCATION + "/*", LIST_BY_SELECTION)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION, MOVIE)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION + "/#", MOVIE_BY_ID)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LOCATION + "/*", MOVIES_BY_SELECTION)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION, MOVIE_LIST)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION + "/#", MOVIE_LIST_BY_ID)
            uriMatcher.addURI(authority, MoviesContract.MOVIE_LIST_LOCATION + "/*", MOVIE_LIST_BY_SELECTION)

            return uriMatcher
        }
    }
}
